skimage/restoration/inpainting.py

```diff
@@ -0,0 +1,183 @@
+    #               2  -8   2
+    #                   1       ]
+    #
+    kernel = [1, 2, -8, 2, 1, -8, 20, -8, 1, 2, -8, 2, 1]
+    offset = [-2 * out_w, -out_w - 1, -out_w, -out_w + 1,
+              -2, -1, 0, 1, 2, out_w - 1, out_w, out_w + 1, 2 * out_w]
+
+    for idx, (i, (m, n)) in enumerate(zip(mask_i, mask_mn)):
+        if 2 <= m <= out_h - 3 and 2 <= n <= out_w - 3:
+            for k, o in zip(kernel, offset):
+                if i + o in mask_i:
+                    matrix_unknown[idx, i + o] = k
+                else:
+                    matrix_known[idx, i + o] = k
+
+    # 2 stage. Find points 1 pixel far from bounds
+    # kernel = [     1
+    #            1  -4  1
+    #                1     ]
+    #
+    kernel = [1, 1, -4, 1, 1]
+    offset = [-out_w, -1, 0, 1, out_w]
+
+    for idx, (i, (m, n)) in enumerate(zip(mask_i, mask_mn)):
+        if (m in [1, out_h - 2] and 1 <= n <= out_h - 2) or \
+           (n in [1, out_w - 2] and 1 <= m <= out_w - 2):
+            for k, o in zip(kernel, offset):
+                if i + o in mask_i:
+                    matrix_unknown[idx, i + o] = k
+                else:
+                    matrix_known[idx, i + o] = k
+
+    # 3 stage. Find points on the horizontal bounds
+    # kernel = [ 1, -2, 1 ]
+    #
+    kernel = [1, -2, 1]
+    offset = [-1, 0, 1]
+
+    for idx, (i, (m, n)) in enumerate(zip(mask_i, mask_mn)):
+        if m in [0, out_h - 1] and 1 <= n <= out_w - 1:
+            for k, o in zip(kernel, offset):
+                if i + o in mask_i:
+                    matrix_unknown[idx, i + o] = k
+                else:
+                    matrix_known[idx, i + o] = k
+
+    # 4 stage. Find points on the vertical bounds
+    # kernel = [  1,
+    #            -2,
+    #             1  ]
+    #
+    kernel = [1, -2, 1]
+    offset = [-out_w, 0, out_w]
+
+    for idx, (i, (m, n)) in enumerate(zip(mask_i, mask_mn)):
+        if n in [0, out_w - 1] and 1 <= m <= out_h - 1:
+            for k, o in zip(kernel, offset):
+                if i + o in mask_i:
+                    matrix_unknown[idx, i + o] = k
+                else:
+                    matrix_known[idx, i + o] = k
+
+    # Prepare diagonal matrix
+    flat_diag_image = sparse.dia_matrix((out.flatten(), np.array([0])),
+                                        shape=(out.size, out.size))
+
+    # Calculate right hand side as a sum of known matrix columns
+    matrix_known = matrix_known.tocsr()
+    rhs = -(matrix_known * flat_diag_image).sum(axis=1)
+
+    # Solve linear system over defect points
+    matrix_unknown = matrix_unknown[:, mask_i]
+    matrix_unknown = sparse.csr_matrix(matrix_unknown)
+    result = spsolve(matrix_unknown, rhs)
+
+    # Handle enormous values
+    out[np.where(out < -1)] = -1
+    out[np.where(out > 1)] = 1
+
+    # Put calculated points into the image
+    for idx, (m, n) in enumerate(mask_mn):
+        out[m, n] = result[idx]
+
+    return skimage.img_as_uint(out)
```

Comment:
The input is double, the output is uint. I have not looked at the algorithm, but is this the way it is supposed to be?